internal/controllers/v1beta2/build_qoshandler.go

```diff
@@ -89,34 +99,76 @@ func (r *LagoonBuildReconciler) processQueue(ctx context.Context, opLog logr.Log
 			}
 			runningNSBuilds, _ := lagooncrd.NamespaceRunningBuilds(pBuild.Namespace, r.BuildCache.Values())
 			// if there are no running builds, check if there are any pending builds that can be started
-			if len(runningNSBuilds) == 0 {
-				opLog.Info("Checking CancelExtraBuilds")
-				if err := lagooncrd.CancelExtraBuilds(ctx, r.Client, opLog, r.QueueCache, r.BuildCache, pBuild.Namespace, "Running"); err != nil {
+			// avoid exceeding total builds
+			if len(runningNSBuilds) == 0 && len(r.BuildCache.Values()) < r.BuildQoS.TotalBuilds {
+				if r.EnableDebug {
+					opLog.Info("Checking UpdateOrCancelExtraBuilds")
+				}
+				startBuild, err := lagooncrd.UpdateOrCancelExtraBuilds(ctx, r.Client, opLog, r.QueueCache, r.BuildCache, pBuild.Namespace)
+				if err != nil {
 					// only return if there is an error doing this operation
 					// continue on otherwise to allow the queued status updater to run
 					return err
+				} else {
+					// start the build pod immediately
+					var lagoonBuild lagooncrd.LagoonBuild
+					if err := r.Get(ctx, types.NamespacedName{Namespace: pBuild.Namespace, Name: startBuild}, &lagoonBuild); err != nil {
+						return err
+					}
+					if err := r.processBuild(ctx, opLog, lagoonBuild); err != nil {
+						// remove it from the cache if there is a failure to process the build
+						r.BuildCache.Remove(lagoonBuild.Name)
+						r.QueueCache.Remove(lagoonBuild.Name)
+					}
 				}
 				// don't handle the queued process for this build, continue to next in the list
 				continue
 			}
 		}
+	}
```

Comment:
So the for that this links to is iterating over the whole of the range of "sortedBuilds", even if there's no work to do - you could, potentially, add a break or something if `idx >= buildsToStart` (or something similar, for instance, only range over a subset of `sortedBuilds` equal in length to `buildsToStart`)
This isn't a train smash, just saw it and thought I'd mention it.